# https://docs.docker.com/compose/compose-file/

name: peak

services:
  web:
    image: caddy:2.9.1-alpine
    restart: unless-stopped
    cap_add:
      - NET_ADMIN
    ports:
      - 8080:8080
      - 8888:8888
      - 2019:2019
    volumes:
      - django_admin_static:/srv/django_static
      - ./web/Caddyfile:/etc/caddy/Caddyfile
      - ./web/:/web/
      - caddy_data:/data
      - caddy_config:/config
  api:
    build: ./backend
    command: python manage.py runserver 0.0.0.0:8000
    volumes:
      - ./backend/:/backend/
      - django_admin_static:/static
    ports:
      - 8000:8000
    depends_on:
      db:
        condition: service_healthy
      redis:
        condition: service_healthy
    env_file:
      - ./.env
  api_cron:
    build: ./backend
    command: crond -f -l 2
    volumes:
      - ./backend/:/backend/
      - ./backend/api.crontab:/etc/crontabs/root
    init: true
    depends_on:
      db:
        condition: service_healthy
      redis:
        condition: service_healthy
    env_file:
      - ./.env
  db: 
    image: postgres:16.2-alpine3.19
    volumes:
      - db_data:/var/lib/postgresql/data/
    healthcheck:
      test: ["CMD", "pg_isready", "-q", "-U", "peakuser", "-d", "peakdb"] 
      # https://github.com/peter-evans/docker-compose-healthcheck/issues/16#issuecomment-1039718816
      interval: 1s
      timeout: 5s
      retries: 10
    env_file:
      - ./.env
  redis:
    container_name: redis-cache
    image: redis
    ports:
      - "6379:6379"
    healthcheck:
      test: ["CMD", "redis-cli", "ping"]
      interval: 1s
      timeout: 5s
      retries: 10
volumes:
  django_admin_static:
  db_data:
    external: true
  caddy_data:
    external: true
  caddy_config: